File: app/src/main/java/com/bopr/android/smailer/util/TextUtil.java
package com.bopr.android.smailer.util;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.bopr.android.smailer.GeoCoordinates;
import com.google.common.base.Function;

import java.util.Collection;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Locale;
import java.util.regex.Pattern;

import static java.lang.String.format;
import static java.lang.String.valueOf;

@SuppressWarnings("WeakerAccess")
public abstract class TextUtil {

    private static final String QUOTED_TEXT_REGEX = "\"([^\"]*)\"";
    private static final String REGEX_ = "REGEX:";
    public static final Pattern QUOTED_TEXT_PATTERN = Pattern.compile(QUOTED_TEXT_REGEX);

    private TextUtil() {
    }

    /**
     * Returns true if string is not empty or null.
     */
    public static boolean isNotEmpty(@Nullable CharSequence s) {
        return !isNullOrEmpty(s);
    }

    /**
     * Returns true if string is empty or null.
     */
    public static boolean isNullOrEmpty(@Nullable CharSequence s) {
        return s == null || s.length() == 0;
    }

    /**
     * Returns true if string filled with spaces or empty or null.
     */
    public static boolean isNullOrBlank(@Nullable String s) {
        return isNullOrEmpty(s) || s.trim().length() == 0;
    }

    @NonNull
    public static String formatCoordinates(@NonNull GeoCoordinates location, String degreeSymbol,
                                           String minuteSymbol, String secondSymbol,
                                           String northSymbol, String southSymbol,
                                           String westSymbol, String eastSymbol) {
        double latitude = location.getLatitude();
        double longitude = location.getLongitude();
        return decimalToDMS(latitude, degreeSymbol, minuteSymbol, secondSymbol)
                + (latitude > 0 ? northSymbol : southSymbol)
                + ", " +
                decimalToDMS(longitude, degreeSymbol, minuteSymbol, secondSymbol)
                + (longitude > 0 ? westSymbol : eastSymbol);
    }

    @NonNull
    public static String formatCoordinates(@NonNull GeoCoordinates location) {
        return formatCoordinates(location, "°", "\'", "\"", "N", "S", "W", "E");
    }

    @NonNull
    public static String decimalToDMS(double coordinate, String degreeSymbol, String minuteSymbol,
                                      String secondSymbol) {
        double crd = coordinate;
        double mod = crd % 1;
        int degrees = Math.abs((int) crd);

        crd = mod * 60;
        mod = crd % 1;
        int minutes = Math.abs((int) crd);

        crd = mod * 60;
        int seconds = Math.abs((int) crd);

        return degrees + degreeSymbol + minutes + minuteSymbol + seconds + secondSymbol;
    }

    @Nullable
    public static String capitalize(@Nullable String text) {
        if (isNullOrEmpty(text)) {
            return text;
        }
        return text.substring(0, 1).toUpperCase(Locale.getDefault()) + text.substring(1);
    }

    @NonNull
    public static String formatDuration(long duration) {
        long seconds = duration / 1000;
        return format(Locale.US, "%d:%02d:%02d", seconds / 3600, (seconds % 3600) / 60, seconds % 60);
    }

    @NonNull
    public static String join(@NonNull Collection values, @NonNull String divider,
                              @Nullable Function<Object, String> transform) {
        StringBuilder sb = new StringBuilder();
        Iterator it = values.iterator();
        while (it.hasNext()) {
            Object value = it.next();
            sb.append(transform != null ? transform.apply(value) : value);
            if (it.hasNext()) {
                sb.append(divider);
            }
        }
        return sb.toString();
    }

    @NonNull
    public static String join(@NonNull Collection values, @NonNull String divider) {
        return join(values, divider, null);
    }

    @NonNull
    public static List<String> split(@NonNull String value, @NonNull String divider) {
        return split(value, divider, null);
    }

    @NonNull
    public static List<String> split(@NonNull String value, @NonNull String divider,
                                     @Nullable Function<String, String> transform) {
        LinkedList<String> list = new LinkedList<>();
        if (!value.isEmpty()) {
            for (String s : value.split(divider)) {
                list.add(transform != null ? transform.apply(s) : s);
            }
        }
        return list;
    }

    @NonNull
    public static String commaJoin(@NonNull Collection values) {
        return join(values, ",", value -> valueOf(value).replaceAll(",", "/,"));
    }

    @SuppressWarnings("ConstantConditions")
    @NonNull
    public static List<String> commaSplit(@NonNull String string) {
        return split(string, "(?<!/),", s -> s.trim().replaceAll("/,", ","));
    }

    @NonNull
    public static String escapeRegex(@NonNull String s) {
        return REGEX_ + s;
    }

    @Nullable
    public static String unescapeRegex(@Nullable String s) {
        if (s != null) {
            int ix = s.indexOf(REGEX_);
            if (ix != -1) {
                return s.substring(ix + REGEX_.length());
            }
        }
        return null;
    }

    public static boolean isQuoted(@Nullable String s) {
        return !isNullOrEmpty(s) && s.charAt(0) == '\"' && s.charAt(s.length() - 1) == '\"';
    }

}

File: app/src/main/kotlin/com/bopr/android/smailer/util/TextUtil.kt
package com.bopr.android.smailer.util

import android.util.Patterns.EMAIL_ADDRESS
import java.io.InputStream
import java.net.URLEncoder
import java.nio.charset.Charset
import java.util.Locale
import java.util.regex.Pattern
import kotlin.math.abs

const val QUOTATION_REGEX = "\"([^\"]*)\""
val QUOTATION_PATTERN: Pattern = Pattern.compile(QUOTATION_REGEX)
val WEB_URL_PATTERN: Pattern = Pattern.compile("(?:\\S+)://\\S+")

private val COMMA_ESCAPED = Regex("(?<!/),")  /* matches commas not preceded by slash symbol */
private const val REGEX_ = "REGEX:"

fun strings(vararg values: Any): Array<String> {
    return Array(values.size) { values[it].toString() }
}

fun escapeRegex(s: String): String {
    return REGEX_ + s
}

fun unescapeRegex(s: String?): String? {
    if (!s.isNullOrEmpty() && s.startsWith(REGEX_)) {
        return s.substring(REGEX_.length)
    }
    return null
}

fun isQuoted(s: String?): Boolean {
    return !s.isNullOrEmpty() && s.first() == '\"' && s.last() == '\"'
}

fun <T> Iterable<T>.commaJoin(): String {
    return joinToString(",") {
        it.toString().replace(",", "/,")  /* escape commas */
    }
}

fun <T> Array<out T>.commaJoin(): String {
    return joinToString(",") {
        it.toString().replace(",", "/,")  /* escape commas */
    }
}

fun String?.commaSplit(): List<String> {
    return if (!isNullOrEmpty()) {
        split(COMMA_ESCAPED).map {
            it.trim().replace("/,", ",")
        }
    } else {
        emptyList() /* to match commaJoin("") */
    }
}

fun decimalToDMS(
    coordinate: Double, degreeSymbol: String, minuteSymbol: String,
    secondSymbol: String
): String {
    var c = coordinate
    var mod = c % 1
    var intPart = c.toInt()

    val degrees = abs(intPart)

    c = mod * 60
    mod = c % 1
    intPart = c.toInt()
    val minutes = abs(intPart)

    c = mod * 60
    intPart = c.toInt()
    val seconds = abs(intPart)

    return degrees.toString() + degreeSymbol + minutes + minuteSymbol + seconds + secondSymbol
}

fun String?.capitalize(): String? {
    return this?.let { substring(0, 1).uppercase() + substring(1) }
}

fun formatDuration(duration: Long?): String? {
    return duration?.let {
        val seconds = duration / 1000
        String.format(Locale.US, "%d:%02d:%02d", seconds / 3600, seconds % 3600 / 60, seconds % 60)
    }
}

fun InputStream.readText(charset: Charset = Charsets.UTF_8): String {
    return bufferedReader(charset).use { it.readText() }
}

fun isValidUrl(url: String?): Boolean {
    return !url.isNullOrBlank() && WEB_URL_PATTERN.matcher(url).matches()
}

fun isValidEmailAddress(address: String?): Boolean {
    return !address.isNullOrBlank() && EMAIL_ADDRESS.matcher(address).matches()
}

fun isValidEmailAddressList(addresses: String?): Boolean {
    return !addresses.isNullOrBlank() && addresses.commaSplit().all { isValidEmailAddress(it) }

}

fun String.htmlReplaceUrlsWithLinks(): String {
    val sb = StringBuffer()

    val matcher = WEB_URL_PATTERN.matcher(this)
    while (matcher.find()) {
        val url = matcher.group()
        matcher.appendReplacement(sb, "<a href=\"$url\">$url</a>")
    }
    matcher.appendTail(sb)

    return sb.toString()
}

fun String.httpEncoded(): String = URLEncoder.encode(this, "UTF-8")
